litestar/openapi/config.py

```diff
@@ -90,31 +108,123 @@ class OpenAPIConfig:
 
     :class:`Reference <litestar.openapi.spec.reference.Reference>` objects.
     """
-    root_schema_site: Literal["redoc", "swagger", "elements", "rapidoc"] = "redoc"
-    """The static schema generator to use for the "root" path of `/schema/`."""
-    enabled_endpoints: set[str] = field(
-        default_factory=lambda: {
-            "redoc",
-            "swagger",
-            "elements",
-            "rapidoc",
-            "openapi.json",
-            "openapi.yaml",
-            "openapi.yml",
-            "oauth2-redirect.html",
-        }
-    )
-    """A set of the enabled documentation sites and schema download endpoints."""
     operation_id_creator: OperationIDCreator = default_operation_id_creator
     """A callable that generates unique operation ids"""
     path: str | None = field(default=None)
-    """Base path for the OpenAPI documentation endpoints."""
+    """Base path for the OpenAPI documentation endpoints.
+
+    If no path is provided the default is ``/schema``.
+
+    Ignored if ``openapi_router`` is provided.
+    """
+    render_plugins: Sequence[OpenAPIRenderPlugin] = field(default=())
+    """Plugins for rendering OpenAPI documentation UIs."""
+    openapi_router: Router | None = None
+    """An optional router for serving OpenAPI documentation and schema files.
+
+    If provided, ``path`` is ignored.
+
+    This parameter is also ignored if the deprecated :class:`OpenAPIConfig <.openapi.OpenAPIConfig>` ``openapi_controller`` kwarg is provided.
+
+    The ``openapi_router`` is not required, but it can be passed to customize the configuration of the router used to serve the documentation endpoints. For example, you can add middleware or guards to the router.
+
+    Handlers to serve the OpenAPI schema and documentation sites are added to this router according
+    to the ``render_plugins`` attribute, so routes shouldn't be added that conflict with these.
+    """
+    openapi_controller: type[OpenAPIController] | None = None
+    """Controller for generating OpenAPI routes.
+
+    Must be subclass of :class:`OpenAPIController <litestar.openapi.controller.OpenAPIController>`.
+    """
+    root_schema_site: Literal["redoc", "swagger", "elements", "rapidoc"] | None = None
```

Comment:
Upon testing some production apps against this it seems to be missing `scalar` unless I'm dumb?

Reply:
In case others come across this, scalar wasn't added to `root_schema_site` literals because that param is deprecated along side the `OpenAPIController` pattern.

To use Scalar in 2.8, one will need to do this:

```py
config = OpenAPIConfig(
    ...
    render_plugins=[ScalarRenderPlugin()]
)
```
Using this pattern implicitly opts the user into "future" mode where `root_schema_site` is ignored and the first plugin named in the `render_plugins` sequence is served at the schema root.

The above pattern (i.e., `ScalarRenderPlugin` as a single render plugin) will become the default behavior in 3.0.

If no `render_plugin` or `openapi_controller` is provided to the config, this enables a compatibility mode, where the future plugin/router based approach is used to emulate the deprecated default `OpenAPIController` behavior.

If an `openapi_controller` is provided, then the user is opted into "classic" mode, which is no change from 2.7.